### DISCUSSION TITLE
[In progress] Add BLE support

### DISCUSSION THREAD

**Discussion by mfucci** [mfucci/node-matter#37] [Workstream]
In Matter, BLE is widely used during the initial commissioning to discover the device and connect it on the WiFi network.

Tasks to do this:

- figure out which BLE library to use (candidates: [noble](https://www.npmjs.com/package/noble) - unmaintained / multi-platform / native code, [node-ble](https://www.npmjs.com/package/node-ble) - maintained / Linux-only / no native code so easier installation)
- create a BLE NetInterface (see UdpInterface.ts) to open a BLE socket
- create a BLE scanner and broadcaster to adviser / discover Matter device
- figure out the advertisement data to advertise that the device can be commissioned
- implement the corresponding fakes to be able to unit test / integration test the system

Contributors:

- @adminy

**Comment by adminy** [comment]
Hi, 

I am interested in helping out. I suggest we use something like web bluetooth right away rather than eventually getting to using it. There is a very well maintained library in deno called [deno-webbluetooth](https://github.com/Symbitic/deno-webbluetooth) and since your library isn't heavily dependent on node specific features I'll see if it works out of the box with deno first. If you have any pointers on some documents/examples I could use to try and extend server / client would be great, otherwise I'm assuming its all in the C++ matter project.

**Comment by mfucci** [comment]
Hi Adminy,

Seems a good idea to use [web bluetooth ](https://developer.mozilla.org/en-US/docs/Web/API/Web_Bluetooth_API) as our abstraction layer to access Bluetooth network, this will make it easier to make node-matter run into a browser!

My recommendations to start on this:

1. Built a Matter device and a Matter controller that successfully communicate over BLE:

  - for the device, a Raspberry Pi should do since this is supported by the official Matter codebase
  - for the controller, either chip-tool in command line or Android ChipTool app. I would recommend to use the Android app since you might have trouble with your computer Bluetooth stack if you are using the command line.
  - BLE sniffer to see the packets. Problably wireshark running on the RPi should be enough
  - make sure that commissioning works over BLE. The device will probably join the WiFi network during the commissioning. Not sure what how to force it to stay on BLE if this is your target. Probably this requires to use Thread protocol

2. Built the device side first and test it with an official Matter controller.

I have already done the initial work to support other communication protocol. 
To implement BLE communication, you need to implement the interfaces: NetInterface.ts, Scanner.ts, Broadcaster.ts

Let me know if you have any trouble or questions!
- Marco

**Reply by adminy** [reply]
Hey Marco,

Thank you for the pointers. I have the nrf52840 dongle on my hands which supports matter but has no wifi, so I could try that and sniff out its packets. I'll let you know how I got on.

**Comment by mfucci** [comment]
As [javierbarellano4v](https://github.com/javierbarellano4v) pointed out on another thread, Matter specs are now publicly available: https://csa-iot.org/developer-resource/specifications-download-request/

This will probably help you figure out how BLE is supported.

**Comment by ilbertt** [comment]
Hello guys, are there any updates on this? I'm really looking forward to use this library for a Node.js-based project :)
